llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp

```diff
@@ -1428,6 +1428,46 @@ static Instruction *foldBitOrderCrossLogicOp(Value *V,
   return nullptr;
 }
 
+/// Fold an unsigned minimum of trailing or leading zero bits counts:
+///   umin(cttz(CtOp, ZeroUndef), ConstOp) --> cttz(CtOp | (1 << ConstOp))
+///   umin(ctlz(CtOp, ZeroUndef), ConstOp) --> ctlz(CtOp | ((1 << (bitwidth-1))
+///                                              >> ConstOp))
+static Value *
+foldMinimumOverTrailingOrLeadingZeroCount(Intrinsic::ID IntrID, Value *I0,
+                                          Value *I1, const DataLayout &DL,
+                                          InstCombiner::BuilderTy &Builder) {
+  assert((IntrID == Intrinsic::cttz || IntrID == Intrinsic::ctlz) &&
+         "This helper only supports cttz and ctlz intrinsics");
+
+  if (I0->hasOneUse()) {
+    if (auto *II0 = dyn_cast<IntrinsicInst>(I0);
+        II0 && II0->getIntrinsicID() == IntrID) {
+      Value *X = II0->getArgOperand(0);
+      Value *Z = II0->getArgOperand(1);
+
+      auto BitWidth = I1->getType()->getScalarSizeInBits();
+      auto *Ty = I1->getType();
+
+      Value *NewCtOp = X;
+      auto LessBitWidth = [BitWidth](auto &C) { return C.ult(BitWidth); };
+      if (match(I1, m_CheckedInt(LessBitWidth))) {
+        Constant *NewConst = ConstantFoldBinaryOpOperands(
+            IntrID == Intrinsic::cttz ? Instruction::Shl : Instruction::LShr,
+            IntrID == Intrinsic::cttz
+                ? ConstantInt::get(Ty, 1)
+                : ConstantInt::get(Ty, APInt::getSignedMinValue(BitWidth)),
+            cast<Constant>(I1), DL);
+        NewCtOp = Builder.CreateOr(X, NewConst);
+      } else if (!match(I1, m_CheckedInt(std::not_fn(LessBitWidth)))) {
+        return nullptr; // Non-splat vector with elements < and >= BitWidth
```

Comment:
Seems a bit unnecessary to essentially re-build the intrip in the `I1 >= BitWidth` case when you just return `I0`. Also would help to preserve attributes.

Reply:
Good point. Done.